llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h

```diff
@@ -2166,6 +2166,9 @@ class OpenMPIRBuilder {
   using TargetBodyGenCallbackTy = function_ref<InsertPointTy(
       InsertPointTy AllocaIP, InsertPointTy CodeGenIP)>;
 
+  using TargetGenArgAccessorsCallbackTy = function_ref<Value *(
+      Argument &Arg, Value *Input, IRBuilderBase &Builder)>;
```

Comment:
Yes, it appears to be necessary, the OpenMPIRBuilder has it's own LLVM builder that it utilises internally which will end up with a different InsertionPoint as the lowering of createTarget progresses, All we currently do is pass an initial insertion point to createTarget and then the internal OpenMPIRBuilder's LLVM builder takes this insertion point and progresses it and then returns the end insertion point back up to the original ModuleTranslation's LLVM builder so it can continue working from wherever the OpenMPIRBuilder left off! That's how I understand it at least. So using the local copy by reference will result in inserting the new instructions at the incorrect location as it won't ever actually be used by the OpenMPIRBuilder.

Reply:
I don't think passing around the builder is the coorect approach to solve this.

We should be passing and returning the inserpoints instead. We should pass the AllocaIP and CodeGenIP to the Callback, and the CB should return the new position of the IP.

If new allocas are not needed the we can perhaps ignore the AllocaIP.

Reply:
I don't think the AllocaIP is currently used, the insertion point that's used at the moment is the first non-phi or debug node, which does result in slightly differing IR to Clang. But I'll utilise the insertion points pass in method and try to get it the IR closer to what Clang currently has.

My only worry of using two differing builders is if they have differing states, but I'm not sure if that is a concern, perhaps the builders only relevant state is the insertion points.